boot/bootutil/include/bootutil/bootutil.h

```diff
@@ -65,6 +65,38 @@ int boot_swap_type(void);
 int boot_set_pending(int permanent);
 int boot_set_confirmed(void);
 
+#define MCUBOOT_REQ_FLASH_MAP_SIZE          0
+#define MCUBOOT_REQ_FLASH_MAP_INFO          1
+
+struct mcuboot_api_itf {
+    uint32_t mcuboot_api_magic;
+    uint32_t mcuboot_version;
+    int (*mcuboot_ioctl)(int req, void *data, size_t len);
+};
+
+extern const struct mcuboot_api_itf mcuboot_api_vt;
+extern struct mcuboot_api_itf *p_mcuboot_api_vt;
+
+#if defined(__arm__)
+    #define MCUBOOT_API_INIT_(m)                                            \
+        extern const struct mcuboot_api_itf mcuboot_api_vt;                 \
+        asm("mov r4, %0\n\t"                                                \
+            "mov r5, %1"  :: "I" (m), "r" (&mcuboot_api_vt) : "r4", "r5");
+#elif defined(__mips__)
+    /* TODO */
+    #define MCUBOOT_API_INIT_(m)
+#elif defined(__i386__)
+    /* TODO */
+    #define MCUBOOT_API_INIT_(m)
+#endif
+
+/* api virtual function call table magic marker */
+#define MCUBOOT_API_MAGIC             0x8b2d7757
+
+/* magic added to a register to mark bootloader api is valid */
+#define MCUBOOT_BOOT_MAGIC            0xb00710ad
+#define MCUBOOT_API_INIT(x)           MCUBOOT_API_INIT_(MCUBOOT_BOOT_MAGIC)
```

Comment:
I find it confusing to leave things in registers after this though. It may work now, but I don't think it is guaranteed to work. I think both the register setup as well as the call are arch specific and should really be together, not split like this.

e.g., other processors may need something other than a call to jump into the app, maybe some explicit instructions and a special kind of jump.